orange/src/main/java/com/hartwig/hmftools/orange/algo/cuppa/CuppaDataFactory.java

```diff
@@ -26,19 +32,101 @@
+                .lineCount(lineCount)
+                .maxComplexSize(maxComplexSize)
+                .simpleDups32To200B(simpleDups32To200B)
+                .telomericSGLs(telomericSGLs)
                 .build();
     }
 
     @NotNull
-    private static List<CuppaPrediction> extractPredictions(@NotNull List<CuppaDataFile> files)
+    public static CuppaData create(@NotNull CuppaPredictions cuppaPredictionsV2)
+    {
+        return create(cuppaPredictionsV2, null);
+    }
+
+    @NotNull
+    public static List<CuppaPrediction> extractSortedProbabilitiesCuppaV2(@NotNull CuppaPredictions cuppaPredictions)
+    {
+        CuppaPredictions probabilitiesAllClassifiers = cuppaPredictions
+                .subsetByDataType(Categories.DataType.PROB)
+                .sortByRank();
+
+        List<CuppaPrediction> cuppaPredictionsOrangeFormat = new ArrayList<>();
+        for(String cancerType : probabilitiesAllClassifiers.CancerTypes)
+        {
+            CuppaPredictions probabilitiesOneCancerType = probabilitiesAllClassifiers.subsetByCancerType(cancerType);
+
+            Map<Categories.ClfName, Double> probabilitiesByClassifier = new HashMap<>();
+            for(int i = 0; i < probabilitiesOneCancerType.size(); i++)
+            {
+                probabilitiesByClassifier.put(
+                        probabilitiesOneCancerType.get(i).ClfName,
+                        probabilitiesOneCancerType.get(i).DataValue
+                );
+            }
+
+            CuppaPrediction prediction = ImmutableCuppaPrediction.builder()
+                    .cuppaMajorVersion("v2")
+                    .cancerType(cancerType)
+                    .likelihood(probabilitiesByClassifier.get(probabilitiesAllClassifiers.MainCombinedClfName))
+                    .genomicPositionClassifier(probabilitiesByClassifier.get(Categories.ClfName.GEN_POS))
+                    .snv96Classifier(probabilitiesByClassifier.get(Categories.ClfName.SNV96))
+                    .eventClassifier(probabilitiesByClassifier.get(Categories.ClfName.EVENT))
+                    .geneExpressionClassifier(probabilitiesByClassifier.get(Categories.ClfName.GENE_EXP))
+                    .altSjClassifier(probabilitiesByClassifier.get(Categories.ClfName.ALT_SJ))
+                    .build();
+
+            cuppaPredictionsOrangeFormat.add(prediction);
+        }
+        return cuppaPredictionsOrangeFormat;
+    }
+
+    @NotNull
+    public static Double getFeatureValue(CuppaPredictions cuppaPredictions, String featureName)
+    {
+        // Feature values are replicated for each cancer type because the `cuppaPredictions` table is in long form.
+        // Use `.findFirst()` to get a single value
+        CuppaPredictionEntry predictionEntry = cuppaPredictions.PredictionEntries.stream()
+                .filter(o -> o.DataType == Categories.DataType.FEAT_CONTRIB & o.FeatName.equals(featureName))
+                .findFirst()
+                .get();
```

Comment:
this isn't great use of java Optional. It would be better to do get().orElseThrow() with an exception with a good message to the user about the problem in data integrity.

Reply:
`.get()` is now replaced with `.orElseThrow(() -> new Exception("Input CuppaPredictions is empty"))`